x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/TopNOperatorTests.java

```diff
@@ -401,12 +600,673 @@ public void testTopNManyDescriptionAndToString() {
+                List.of(5000L, 1000L, 7000L)
+            )
+        );
+        expectedValues.add(null);
+        expectedValues.add(null);
+
+        assertSortingOnMV(
+            LONG_MV,
+            expectedValues,
+            LONG,
+            new TopNOperator.SortOrder(0, true, false),
+            new TopNOperator.SortOrder(0, false, false)
+        );
+    }
+
+    public void testTopNWithSortingOnSameField_ASC_then_DESC_double() {
+        List<Object> expectedValues = new ArrayList<>();
+        expectedValues.addAll(
+            List.of(
+                List.of(63d, -61d, -62.123d),
+                List.of(1d, 5d, -1d, -0.01d, -5d),
+                List.of(-1d, -0.01d, 1d, 0.01d),
+                List.of(0d, 0d, 17d),
+                List.of(0.01d, 0.01d, 1d, 1d, 1d),
+                100d
+            )
+        );
+        expectedValues.add(null);
+
+        assertSortingOnMV(
+            DOUBLE_MV,
+            expectedValues,
+            DOUBLE,
+            new TopNOperator.SortOrder(0, true, false),
+            new TopNOperator.SortOrder(0, false, false)
+        );
+    }
+
+    public void testTopNWithSortingOnSameField_ASC_then_DESC_BytesRef() {
+        assertSortingOnMV(
+            INT_MV,
+            List.of(
+                List.of(new BytesRef("-1"), new BytesRef("63"), new BytesRef("2")),
+                new BytesRef("100"),
+                List.of(new BytesRef("22"), new BytesRef("21"), new BytesRef("22")),
+                new BytesRef("50"),
+                List.of(new BytesRef("63"), new BytesRef("61"), new BytesRef("62"))
+            ),
+            BYTES_REF,
+            new TopNOperator.SortOrder(0, true, false, BYTES_REF_ENCODER),
+            new TopNOperator.SortOrder(0, false, false, BYTES_REF_ENCODER)
+        );
+    }
+
+    private void assertSortingOnMV(
+        List<List<Object>> values,
+        List<Object> expectedValues,
+        ElementType blockType,
+        TopNOperator.SortOrder... sortOrders
+    ) {
+        Block block = TestBlockBuilder.blockFromValues(values, blockType);
+        assert block.mvOrdering() == Block.MvOrdering.UNORDERED : "Blocks created for this test must have unordered multi-values";
+        Page page = new Page(block);
+
+        List<List<Object>> actualValues = new ArrayList<>();
+        int topCount = randomIntBetween(1, values.size());
+        try (
+            Driver driver = new Driver(
+                new DriverContext(),
+                new CannedSourceOperator(List.of(page).iterator()),
+                List.of(new TopNOperator(topCount, List.of(sortOrders), randomPageSize())),
+                new PageConsumerOperator(p -> readInto(actualValues, p)),
+                () -> {}
+            )
+        ) {
+            runDriver(driver);
+        }
+        assertMap(actualValues, matchesList(List.of(expectedValues.subList(0, topCount))));
+    }
+
+    public void testRandomMultiValuesTopN() {
```

Comment:
The meat of the TopN tests that randomizes the number of blocks, of rows and the blocks to sort by.